java/code/src/com/redhat/rhn/manager/action/ActionManager.java

```diff
@@ -2344,4 +2367,86 @@ public static ImageInspectAction scheduleImageInspect(User scheduler, List<Long>
         scheduleForExecution(action, new HashSet<>(sids));
         return action;
     }
+
+    /**
+     * Connect given systems to another proxy.
+     *
+     * @param loggedInUser The current user
+     * @param sysids A list of systems ids
+     * @param proxyId Id of the proxy or 0 for direct connection to SUMA server
+     * @return Returns a list of scheduled action ids
+     *
+     */
+
+    public static List<Long> changeProxy(User loggedInUser, List<Long> sysids, Long proxyId)
+        throws TaskomaticApiException {
+        List<Long> visible = MinionServerFactory.lookupVisibleToUser(loggedInUser)
+                    .map(m -> m.getId()).collect(toList());
+        if (!visible.containsAll(sysids)) {
+            sysids.removeAll(visible);
+            throw new UnsupportedOperationException("Some System not available or not managed with Salt: " + sysids);
+        }
+
+        List<MinionServer> minions = sysids.stream().map(
+            id -> SystemManager.lookupByIdAndUser(id, loggedInUser).asMinionServer().get()).collect(toList());
+
+        List<Long> proxies = minions.stream().filter(m -> m.isProxy()).map(m -> m.getId()).collect(toList());
+        if (!proxies.isEmpty()) {
+            throw new UnsupportedOperationException("Some of the minions are proxies: " + proxies);
+        }
```

Comment:
Does this mean that second level proxies cannot be moved under different proxy?

Reply:
Yes. Moving a proxy is much more difficult. The code would have to handle traditional stack, cycles, etc.

Reply:
Understood, this needs to be documented.